src/controls/keyboard.py

```diff
@@ -0,0 +1,93 @@
+from pynput.keyboard import Key, Listener
+from connection import connecter, Topic
+import os
+import sys
+dir_path = os.path.dirname(os.path.realpath(__file__))
+parent_dir_path = os.path.abspath(os.path.join(dir_path, os.pardir))
+sys.path.insert(0, parent_dir_path)
+
+IDLE = 0
+STRAIGHT = 0
+VELOCITY = 100
+DIRECTION = 100
+MQTT_VELOCITY_FORWARD = connecter.create("Forward")
+MQTT_VELOCITY_BACKWARD = connecter.create("Backward")
+MQTT_STEERING = connecter.create("steering")
+TOPIC = Topic()
+
+
+class Keyboard:
+    def __init__(self):
+        self.up = False
+        self.down = False
+        self.left = False
+        self.right = False
+        self.crash_expected = False
+        self.front_sensor = connecter.create("front-sensor")
+        self.front_sensor.subscribe(TOPIC.front_sensor)
+        self.front_sensor.on_message = self.on_message_front_sensor
+        self.front_sensor.loop_start()
+        self.infrared_front = connecter.create("infrared-front")
+        self.infrared_front.subscribe(TOPIC.infrared_front)
+        self.infrared_front.on_message = self.on_message_infrared_front
+        self.infrared_front.loop_start()
+        self.listener = Listener(
+            on_press=self.on_press, on_release=self.on_release)
+        self.listener.start()
+
+    def on_message_front_sensor(self, client, userdata, msg):
+        print(f"ultrasonic - {msg.payload}")
+        if int(msg.payload) > 200 and self.up is True:
+            MQTT_VELOCITY_FORWARD.publish(TOPIC.throttle_forward, IDLE)
+            self.crash_expected = True
+        else:
+            self.crash_expected = False
+
+    def on_message_infrared_front(self, client, userdata, msg):
+        print(f"infrared - {msg.payload}")
+        if int(msg.payload) > 0 and self.up is True:
+            MQTT_VELOCITY_FORWARD.publish(TOPIC.throttle_forward, IDLE)
+            self.crash_expected = True
+        else:
+            self.crash_expected = False
```

Comment:
I don't get this, if you receive a distance **more** than 2 meters then you stop and when you receive an infrared distance that's not 0 you throttle?

Reply:
The simple implementation we were going for was that if the sensor read anything at all, other than zero, we would stop it running. 
Instead of this I have adapted the smartcar.ino file to apply the obstacle avoidance, and incorporated logic that works off a range instead. @DrakeAxelrod and @MrJonte are adding some more logic now and will push the code a bit later today.